### ISSUE TITLE
Request: Expand the event syntax to support modifying ships and outfits

### ISSUE THREAD

**Issue by Zitchas** [endless-sky/endless-sky#5756]
**Is your feature request related to a problem? Please describe.**
Going forward in the Remnant campaign, the player will be introduced to the fact that the Remnant are exploiting two of the big advantages of having a tightly integrated Research & Development, Production, and End User ecosystem: Namely, "rapid iteration" and "active testing & feedback by the end user." 

This leads to situations where in the course of testing over a series of missions the Remnant are using several different versions of an outfit, potentially across a range of different ships and fleets. If an attempt at realistic time frames is implemented, this leads to a situation where an outfit is rolled out for use, used as stock equipment on a range of ships for a few months, then additional missions (or just events) that result in a new version of the outfit being rolled out to those ships, repeat until the satisfactory result is achieved. 

Under the current system, each iteration of the outfit will require a separate outfit. Then there will need to be a new variant of each ship that utilizes that version of the outfit. And then new fleets that include the new variant of the ship. And for outfits that could be reasonably assumed to be widely used across many different variants, there will need to be new variants of each variant. 

As such, the amount of changes required cascade rapidly for even a small series of outfit changes.

**Related Issue Links**
Couldn't see any, but will look more after.

**Describe the solution you'd like**
Have the `event` syntax support changes to outfit and ship data types. Ideally, this would include:
- The ability to remove an attribute from a ship or outfit
- The ability to add an attribute to a ship or outfit
- The ability to modify an existing attribute to a ship or outfit.
- The ability to add or remove outfits on a ship. 

For the most part, I am considering these in respects to the stock ships and outfits. However, if it was possible to directly apply these changes to the player's flagship instead of to the specific ship listed in the mission, that'd be neat too.

**Describe alternatives you've considered**
Simply using multiple different outfits to represent the series of changes. This necessitates different ship variants to cover each time the outfit changes. Which in turn requires changes to fleets to incorporate those changes.

In the case of situations where there may be multiple different missions changing ships, the number of required variants to cover all possibilities cascades exponentially.

**Additional context**
Here is the rough summary of something I'd like to be able to do. Implementation, of course, is subject to change until the content is actually made:

> As a simple version of one thing that I am planning, I want to have the Penguin mature over time.
> 1. So we start with the base penguin we know and love.
> 2. At some point after the player has unlocked the research laboratory, Taely figures out how to incorporate it into a penguin (penguins gain +6 space and come default equipped with a Research Laboratory.)
> 3. After technological improvements a bit later in the cognizance line, the player has the choice of suggesting the Penguin should be encouraged to be tougher, or to have better endurance. (former gives increase to shields, hull, and outfit space; later gives increase to bunks, cargo, fuel, and a smaller boost to outfit space. (we're talking small amounts here.)
> 4. Later chapters will have other improvements. For instance Should the Penguin be better at attacking, or better to have a defensive weapon emplacement? (former gives second gun port, later gives a restricted angle rear-facing turret hardpoint. Both would give a bit of extra weapon space)
> 
> All of these, however, are side missions with extra conditions and not direct parts of the main story. So some people will get some of those, others will get others of them, and obviously people will make different choices. If I'm not mixing up my math, in the current implementation, just those four steps will require at least 8 different penguin ship definitions. There's a *lot* more chapters and side strings coming, and at least a few of them will have their own little tie-ins to the Penguin. Storywise, I literally want the Penguin to be <first>'s pet. (if they want it, anyway). Another of those changes that I'm thinking of is having the choice to either have it redesignated as a fighter (so it can be carried) or have better ramscoop abilities for better range. Maybe have it be reclassified as an escape pod for the flagship once that's merged.
> There is also the likelyhood that while the player's penguin is growing up alongside them; penguins available in the shipyard are younger and not yet capable of most instances of such growth. As such, they would only get some of these improvements, but not most of them.

**Comment by tehhowch** [comment]
Enabling these extensions removes a significant strength in the game's reference-checking functionality, as we can no longer assume that outfits or ships that are missing on initial data parse are actually missing / ill-formed, nor can we ever reason about whether a stock ship model always obeys the limits on outfit/weapon/engine space.

To maintain those guarantees, we need to exhaustively determine the viable permutations of definitions and evaluate each one. That's not functionality that belongs in the game, but in a separate data analysis tool, which means only the most determined content authors will even try to use it.

The fungibility of outfits (a single instance exists and everything references it) means that allowing changes via events instantly modifies every use of the outfit, however nonsensical ("my ship parked in Korath space that has never seen an outfitter now has all these extra tech stats!"). In that light, it would not make sense to allow an event to modify any functional or observable qualities of an outfit, which means events might as well never modify outfits to begin with.

I'm more amenable to modifying stock ship definitions, since these changes will only be available to new ship instances and not existing npcs/player ships.

**Comment by Amazinite** [comment]
As has been mentioned before, you can already swap out ship variants through events, which would gameplay wise be the same thing as changing a ship definition. As far as "upgrading" outfits through events goes, one could simply define separate outfits whos names indicate that they are newer versions of the same outfit (e.g. "Energy Cannon MKII"), or we could look into having outfit variants like we have ship variants (#927), making the upgrade more seamless without affecting already installed outfits.

**Comment by Zavindur** [comment]
tldr;
Maybe allow missions to take a ship from the player?

Longer Version:
I've thought it would be neat to allow missions to directly change things dynamically for example, to make a weapon or outfit gets improved specs; how it is now- I would need to pre-generate all variations of the outfit; so I couldn't just say +5 dps, I would need to have lots of conditional settings to figure which outfit they are currently on to determine which one they should go to next.

I think this would be more doable if the saved files actually kept a copy of each outfit info separately to prevent updating outfits on your ships on the other side of the galaxy; but I'm not a fan of complicating things;

With having the remnant ship 'grow'; is there a way to take a ship from the player?  I know we can give ships to the player; but it may work out if there was a way to have a mission where you need to leave the ship at some facility (and dump all the outfits into the player's cargo).  Then X time later they get "their ship" back all nice and upgraded.

If they can by multiple copies of the ship, you may need some way to track 'how old' a ship is, maybe access to the ship hull resell value would do that.

**Comment by Zitchas** [comment]
I have requested the "remove ship" feature, but there are a few things that are needed before we get that. But yes, people are working on it.

**Comment by Amazinite** [comment]
For the reasons that tehhowch described, I don't think it makes sense to have events modify ships and outfits. As I mentioned, you can already create "upgraded" ships through the use of variants, and although we don't have outfit variants, we can give outfits display names. Providing such upgrades to a player would then just be a matter of making the ships and/or outfits available for them to swap out. Missions were already able to take outfits and at this point can also take ships. A feature like #4217 is all that would be needed to seamlessly swap out every item that a player has for the upgraded version.
